src/codegen/toplevel_definitions.jl

```diff
@@ -167,9 +167,27 @@ function codegen(io, t::EnumType, ::Context)
     maybe_generate_reserved_fields_method(io, t)
 end
 
-function codegen(io, t::ServiceType, ::Context)
-    println(io, "# TODO: SERVICE")
-    println(io, "#    ", safename(t))
+
+const _service_codegen_handlers = Vector{Function}()
+
+"""
+    register_service_codegen(f::Function)
+
+Register a function that will be called for each service in order to generate client/server code stubs.
+The function `f` should have the signature `f(io, t::ProtoBuf.CodeGenerators.ServiceType, ::ProtoBuf.CodeGenerators.Context)`
+"""
+register_service_codegen(f::Function) = push!(_service_codegen_handlers, f)
```

Comment:
/
```suggestion
function register_service_codegen(f::Function) 
    if !(f in _service_codegen_handlers)
        push!(_service_codegen_handlers, f)
    end
end
```
I was calling this from a script and wondered why I suddenly had 20 copies of the same RPCs 😄

Reply:
I resolved this by using a Dict instead which lets us automatically manage the uniqueness of the handler.